pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -992,38 +994,34 @@ func (r *ReconcileClusterDeployment) fixupAdminKubeconfigSecret(secret *corev1.S
 }
 
 // setAdminKubeconfigStatus sets all cluster status fields that depend on the admin kubeconfig.
-func (r *ReconcileClusterDeployment) setAdminKubeconfigStatus(cd *hivev1.ClusterDeployment, adminKubeconfigSecret *corev1.Secret, cdLog log.FieldLogger) error {
-	if cd.Status.WebConsoleURL == "" || cd.Status.APIURL == "" {
-		remoteClusterAPIClient, err := r.remoteClusterAPIClientBuilder(string(adminKubeconfigSecret.Data[adminKubeconfigKey]), controllerName)
-		if err != nil {
-			cdLog.WithError(err).Error("error building remote cluster-api client connection")
-			return err
-		}
-
-		// Parse the admin kubeconfig for the server URL:
-		config, err := clientcmd.Load(adminKubeconfigSecret.Data["kubeconfig"])
-		if err != nil {
-			return err
-		}
-		cluster, ok := config.Clusters[cd.Spec.ClusterName]
-		if !ok {
-			return fmt.Errorf("error parsing admin kubeconfig secret data")
-		}
+func (r *ReconcileClusterDeployment) setAdminKubeconfigStatus(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) error {
+	if cd.Status.WebConsoleURL != "" || cd.Status.APIURL != "" {
+		return nil
+	}
 
-		// We should be able to assume only one cluster in here:
-		server := cluster.Server
-		cdLog.Debugf("found cluster API URL in kubeconfig: %s", server)
-		cd.Status.APIURL = server
-		routeObject := &routev1.Route{}
-		err = remoteClusterAPIClient.Get(context.Background(),
-			types.NamespacedName{Namespace: "openshift-console", Name: "console"}, routeObject)
-		if err != nil {
-			cdLog.WithError(err).Error("error fetching remote route object")
-			return err
-		}
-		cdLog.Debugf("read remote route object: %s", routeObject)
-		cd.Status.WebConsoleURL = "https://" + routeObject.Spec.Host
+	// The secondary API URL is used here because we always want to use the initial API URL immediately after installation.
+	remoteClientBuilder := r.remoteClusterAPIClientBuilder(cd).UseSecondaryAPIURL()
```

Comment:
This comment and code had me confused initially, I assumed SecondaryAPIURL meant OverrideAPIURL, I see that's not the case. The comment could possibly use a little clarification but the code is probably about as clear as it can be.

Are we guaranteed that at this point, the original API URL is still functional and no operators could have replaced it yet? It seems unlikely, but is it guaranteed?

Reply:
In all but a restore scenario, we are guaranteed that Hive has not laid down anything that would make the original API URL no longer accessible. At this point, the .spec.installed field is still false, so no other controllers should be trying to communicate with the remote cluster.

That being said, I think it will be better to try both URLs to cover any scenarios where one may be inaccessible.

Reply:
I reworked the conditions so that the clsuterdeployment controller does not need to explicitly use one of the URLs. The controller will now use the active URL.